timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/timestamp/TimestampRangesTest.java

```diff
@@ -0,0 +1,86 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.transaction.timestamp;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.Test;
+
+import com.palantir.timestamp.TimestampRange;
+
+public class TimestampRangesTest {
+    private static final long SEVENTY_THREE = 73L;
+    private static final long EIGHTY_TWO = 82L;
+
+    private static final TimestampRange SEVENTY_THREE_TO_EIGHTY_TWO
+            = TimestampRange.createInclusiveRange(SEVENTY_THREE, EIGHTY_TWO);
+
+    @Test
+    public void canGetTimestampFromRangeIfItIsTheLowerBound() {
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 3, 10))
+                .isPresent()
+                .hasValue(SEVENTY_THREE);
+    }
+
+    @Test
+    public void canGetTimestampFromRangeIfItIsTheUpperBound() {
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 2, 10))
+                .isPresent()
+                .hasValue(EIGHTY_TWO);
+    }
+
+    @Test
+    public void canGetTimestampsFromRangeInTheMiddle() {
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 7, 10))
+                .isPresent()
+                .hasValue(77L);
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 8, 10))
+                .isPresent()
+                .hasValue(78L);
+    }
+
+    @Test
+    public void canHandleMultipleValidMatches() {
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 1, 2))
+                .isPresent()
+                .satisfies(optionalLong -> {
+                    long value = optionalLong.getAsLong();
+                    assertThat(value).isIn(73L, 75L, 77L, 79L, 81L);
+                });
+        assertThat(TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 0, 2))
+                .isPresent()
+                .satisfies(optionalLong -> {
+                    long value = optionalLong.getAsLong();
+                    assertThat(value).isIn(74L, 76L, 78L, 80L, 82L);
+                });
+    }
+
+    @Test
+    public void throwsIfResidueExceedsOrEqualsModulus() {
+        assertThatThrownBy(() -> TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 2, 2))
+                .isInstanceOf(IllegalArgumentException.class);
+        assertThatThrownBy(() -> TimestampRanges.getTimestampMatchingModulus(SEVENTY_THREE_TO_EIGHTY_TWO, 3, 2))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void throwsIfTimestampRangeDoesNotContainAnyValuesMatchingModulus() {
+        TimestampRange oneTimestamp = TimestampRange.createInclusiveRange(77, 77);
+        assertThat(TimestampRanges.getTimestampMatchingModulus(oneTimestamp, 6, 10)).isNotPresent();
```

Comment:
this doesn't check that we're throwing?

Reply:
Right. This is an artifact of when I previously had the API throw if nothing was found - before I realised that there are legitimate cases where `timestampService.getFreshTimestamps(N)` might actually return you nothing that matches a given residue modulo N - will rename the test.